radix-transaction-scenarios/src/scenarios/all_scenarios.rs

```diff
@@ -1,54 +1,422 @@
+            bootstrap: self.bootstrap,
+            starting_nonce: self.starting_nonce,
+            nonce_handling: self.nonce_handling,
+            network_definition: self.network_definition,
+            /* Callbacks */
+            on_transaction_execution: self.on_transaction_execution,
+            on_scenario_start: self.on_scenario_start,
+        }
+    }
+
+    /// Sets the Native VM to use for the scenarios execution.
+    pub fn native_vm<NE: NativeVmExtension>(
+        self,
+        native_vm: NativeVm<NE>,
+    ) -> TransactionScenarioExecutor<D, W, NE, F1, F2> {
+        TransactionScenarioExecutor {
+            /* Environment */
+            database: self.database,
+            scrypto_vm: self.scrypto_vm,
+            native_vm,
+            /* Execution */
+            scenarios_to_execute: self.scenarios_to_execute,
+            bootstrap: self.bootstrap,
+            starting_nonce: self.starting_nonce,
+            nonce_handling: self.nonce_handling,
+            network_definition: self.network_definition,
+            /* Callbacks */
+            on_transaction_execution: self.on_transaction_execution,
+            on_scenario_start: self.on_scenario_start,
+        }
+    }
+
+    /// The scenarios that the executor should execute. They are not specified per scenario but per
+    /// requirement. Meaning that through this field a client may opt-in or opt-out of the scenarios
+    /// of an entire protocol update but not individual ones.
+    pub fn scenarios_to_execute(
+        mut self,
+        scenario_requirements: impl IntoIterator<Item = ScenarioRequirements>,
+    ) -> Self {
+        self.scenarios_to_execute = scenario_requirements.into_iter().collect();
+        self
+    }
+
+    /// Controls whether the bootstrap process should be performed or not.
+    pub fn bootstrap(mut self, bootstrap: bool) -> Self {
+        self.bootstrap = bootstrap;
+        self
+    }
+
+    /// Sets the starting nonce for executing scenarios.
+    pub fn starting_nonce(mut self, starting_nonce: u32) -> Self {
+        self.starting_nonce = starting_nonce;
+        self
+    }
+
+    /// Defines how the executor should handle nonces.
+    pub fn nonce_handling(mut self, nonce_handling: NonceHandling) -> Self {
+        self.nonce_handling = nonce_handling;
+        self
+    }
+
+    /// Sets the network definition to use for the scenario execution.
+    pub fn network_definition(mut self, network_definition: NetworkDefinition) -> Self {
+        self.network_definition = network_definition;
+        self
+    }
+
+    /// Sets the callback to call after executing a scenario transaction.
+    pub fn on_transaction_execution<
+        NF1: FnMut(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
+    >(
+        self,
+        callback: NF1,
+    ) -> TransactionScenarioExecutor<D, W, E, NF1, F2> {
+        TransactionScenarioExecutor {
+            /* Environment */
```

Comment:
For all of these you should be able to return this instead:
```rust
    Self {
        on_transaction_execution: callback,
        ..self,
    }
```

Or perhaps even simpler: 
```rust
pub fn on_transaction_execution<
        NF1: FnMut(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
    >(mut self, callback: NF1) -> Self {
    self.on_transaction_execution = callback;
    self
}
```

Reply:
Ahh well then I think it's better to lose the generics and just box the callbacks.

Reply:
Happy to pair with you on the refactor :D

Reply:
Hey! I've made the change to using dynamic dispatch on this (and commented in another comment/thread, but commenting here :D)

I've tried switching this to use dynamic dispatch and it works really nice in this context, I've typically been quite reserved about using dynamic dispatch especially in contexts where the compiler can tell what the types are with some extra help. However, I think that this builder (and probably others that have tons of generics) benefited well (syntactically) from using dynamic dispatch.

It still feels like using generics might be better, but it's not nearly as maintainable, and is quite hard to write. It seems like this is a perfect area to use macros to define these [builders with type state](https://www.youtube.com/watch?v=k8kd22jNcps).